pkg/rules/queryable.go

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package rules
+
+import (
+	"context"
+	"math/rand"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/querier/series"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/thanos-io/thanos/pkg/httpconfig"
+	"github.com/thanos-io/thanos/pkg/promclient"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+type promClientsQueryable struct {
+	httpMethod string
+	step       time.Duration
+
+	logger              log.Logger
+	promClients         []*promclient.Client
+	queryClients        []*httpconfig.Client
+	restoreIgnoreLabels []string
+
+	duplicatedQuery prometheus.Counter
+}
+type promClientsQuerier struct {
+	ctx        context.Context
+	mint, maxt int64
+	step       int64
+	httpMethod string
+
+	logger              log.Logger
+	promClients         []*promclient.Client
+	queryClients        []*httpconfig.Client
+	restoreIgnoreLabels []string
```

Comment:
Perhaps we could make this a bit more generic and rename it to `ignoreLabelNames`?

Reply:
Updated.